Scavenger

ShobanaSekar opened the pull request.

dhml commented:
@ShobanaSekar, I would suggest opening a ticket with RTD.  The branch continuous deployment automation and underlying infrastructure are completely out of our hands as its a hosted service.

ShobanaSekar commented:
@gshomo @manasvita @marianalevineu Please review and let me know your comments/suggestions, thank you!

dhml commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+
+Introduction
+===================
+‘Scavenger’ is a new partition on discovery that allows the research community to use private resources 
```
Capitalize Discovery here.

dhml commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+on the cluster when they are idle. ‘Scavenger’ has hardware not otherwise available to the general research 
+community and, in time, will double the resources to which users have access.
+
+When to use scavenger
```
Capitalize Scavenger here, although it will be specified as `scavenger` (lower-case) in configuration and commands I gather.

dhml commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+When to use scavenger
+===================
+
+From best to less-good case scenarios:
```
"least" instead of "less"?

dhml commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+we will exclude partition ``short`` and focus on ``scavenger``. When you run the ``squeue`` command, you can see 
+that your job has been assigned to the scavenger partition::
+
+  [m.joshi@login-01 ~]$ srun  -p scavenger --pty /bin/bash
```
Superfluous extra space between `srun` and `-p` here can be removed.

dhml commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+
+For the example provided on our `checkpointing <https://rc-docs.northeastern.edu/en/latest/best-practices/checkpointing.html?highlight=array#gromacs-checkpointing-example>`_ page, you can use the scavenger partition as::
+
+ #!/bin/bash
```
Can we block-quote this script so it appears as such?

ShobanaSekar replied:
@dhml It is currently block quoted in a way. You can see how it looks like here https://northeastern-university-rc-public-documentation--19.com.readthedocs.build/en/19/ and navigating to 'Hardware on Discovery' > Scavenger (The link will be under 'All checks have passed' > Show all checks > Details if you scroll down on the PR page). Is this what you meant?

dhml replied:
@ShobanaSekar, look good and thanks for the pointer!  I'm not yet that familiar with Sphinx and `*.rst` files.

dhml commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+===================
+
+If labs have purchased a partition, they have priority access to those resources. 
+In the case of scavenger, this means that if your job is running on a private partition 
```
Capitalize Scavenger here.

dhml commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+If labs have purchased a partition, they have priority access to those resources. 
+In the case of scavenger, this means that if your job is running on a private partition 
+from a specific PI’s lab that was idle but is now necessary for a job from that PI’s lab, 
+your job will be killed and re-queued. If you use checkpointing, this would be less of an issue. 
```
Is the plan to automatically re-queue killed jobs?

ShobanaSekar replied:
I don't believe so. @gshomo could you clarify?

ShobanaSekar replied:
I thought so but now I am not sure, @gshomo could you clarify?

dhml replied:
Per discussion in AMA yesterday, jobs will be re-queued.

dhml commented:
@ShobanaSekar, all of the `.DS_Store` files should be removed.

gshomo reviewed (approved):
looks good to me

marianalevineu commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+we will exclude partition ``short`` and focus on ``scavenger``. When you run the ``squeue`` command, you can see 
+that your job has been assigned to the scavenger partition::
+
+  [m.joshi@login-01 ~]$ srun -p scavenger --pty /bin/bash
```
or replacing `m.joshi` with a generic `username`

manasvita replied:
The entire square bracket content can be taken off. I don't think we show the terminal prompt content in any of our documentation. All srun commands are in code block anyway.

marianalevineu replied:
@manasvita  agreed, @ShobanaSekar  can you change that to remove the terminal prompt?

marianalevineu commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+If labs have purchased a partition, they have priority access to those resources. 
+In the case of Scavenger, this means that if your job is running on a private partition 
+from a specific PI’s lab that was idle but is now necessary for a job from that PI’s lab, 
+your job will be killed and re-queued. If you use checkpointing, this would be less of an issue. 
```
might worth mentioning it will be killed within a time window, I think it's 90 seconds but I might be wrong we can double check with @gshomo. It's also worth saying the it's re-queued and there's no guarantee the job will be up and running right away.

manasvita replied:
It would be good to edit lines 66-68 to get the message across more clearly. So something like:
"Members associated with a private partition have priority access to resources attached with that partition. In the case of Scavenger, this means that if a non-member's job is running on a private partition and a job request from a member comes in at that time then the non-member's job will be killed and re-queued."

ShobanaSekar replied:
@marianalevineu I've mentioned it is automatically killed and requeued and also checked with Greg on the time fram but it seems to be a bit ambiguous to me, he said it just depends on the slurm config and could be 30s or so. So I am not mentioning a number here if thats okay?

marianalevineu replied:
@ShobanaSekar  I think that's fine, thanks for checking.

marianalevineu reviewed (changes requested):
Hey @ShobanaSekar  I've reviewed the scavenger page, overall looks good. You can address my suggestions when you get the chance. thanks.

manasvita commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+When to use Scavenger
+===================
+
+From best to least-good case scenarios:
```
Maybe, change it to "From most- to least-desirable case scenarios:"

manasvita commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,69 @@
+Introduction
+===================
+‘Scavenger’ is a new partition on Discovery that allows the research community to use private resources 
+on the cluster when they are idle. ‘Scavenger’ has hardware not otherwise available to the general research 
```
Maybe you can say: "allows the research community to use resources associated with a private partition when they are idle"

ShobanaSekar commented:
@dhml @manasvita @marianalevineu I've addressed most comments, in case one of you wanted to take another quick look. I think this is ready to be merged otherwise. Thanks!

marianalevineu reviewed (changes requested):
Hey @ShobanaSekar I've added a couple of final remarks, other than that I think it looks good. Please address those when you get a chance thanks!

dhml commented:
@ShobanaSekar, it appears that several of my previous comments haven't been addressed yet.  They're minor things, but best to reply to those and resolve them before merging.  Thanks...

ShobanaSekar commented:
> @ShobanaSekar, it appears that several of my previous comments haven't been addressed yet. They're minor things, but best to reply to those and resolve them before merging. Thanks...

@dhml I think I've addressed your comments, but its not reflecting in the rtd page. The browser cache usually needs to be cleared out or sometimes the changes won't reflect. So to see the modified page you'd first have to clear cache and then under the PR page, go to 'All checks have passed' > Show all checks > Click on 'Details'. I ran into the same issue as well and got resolved after clearing cache. Let me know if that works.

dhml commented:
@ShobanaSekar, right thanks, indeed I can see the changes were actually made, thanks!  What I mean is we should reply to and resolve the conversation threads once items have been addressed.  This both provides confirmation to the commenter and simplifies the conversation view to only contain unresolved items so the author and reviewers can quickly see what items are outstanding.

dhml commented:
> @ShobanaSekar, all of the `.DS_Store` files should be removed.

@ShobanaSekar, it looks like this one was missed.  There are still two such files in the PR.  These files are metadata specific to Mac OS and shouldn't be source-controlled.

ShobanaSekar commented:
> @ShobanaSekar, right thanks, indeed I can see the changes were actually made. What I mean is we should reply to and resolve the conversation threads once items have been addressed. This both provides confirmation to the commenter and simplifies the conversation view to only contain unresolved items so the author and reviewers can quickly see what items are outstanding. I've resolved all of these threads.

Sounds good, thanks @dhml

ShobanaSekar commented:
> > @ShobanaSekar, all of the `.DS_Store` files should be removed.
> 
> @ShobanaSekar, it looks like this one was missed. There are still two such files in the PR. These files are metadata specific to Mac OS and shouldn't be source-controlled.

Thanks @dhml, this should be removed now.

ShobanaSekar commented on docs/source/hardware/scavenger.rst:
```diff
@@ -0,0 +1,70 @@
+Introduction
+===================
+‘Scavenger’ is a new partition on Discovery that allows the research community to use resources associated with a 
+private partition on the cluster when they are idle. ‘Scavenger’ has hardware not otherwise available to the general research 
```
This is done now

ShobanaSekar commented:
@marianalevineu have addressed your second set of comments, let me know if there was anything else. I think this should be good to go now, thanks!

dhml reviewed (approved):
@ShobanaSekar, looks good, thanks!

marianalevineu reviewed (approved):
@ShobanaSekar  thanks looks good, I approved this PR.

dhml commented:
> @ShobanaSekar, all of the `.DS_Store` files should be removed.

It appears the `.DS_Store` files are still present.

ShobanaSekar commented:
> > @ShobanaSekar, all of the `.DS_Store` files should be removed.
> 
> It appears the `.DS_Store` files are still present.

Thanks @dhml, these should be removed now. I think they are getting added each time I commit/push.

ShobanaSekar commented:
> the main file needs to be renamed to lowpriority.rst, also should be updated in index.rst: hardware/scavenger -> hardware/lowpriority

Thanks @rschroter, should be done now.

ShobanaSekar commented:
@dhml @manasvita @marianalevineu Re-requesting this review since the language has been modified to be consistent with the FAQs [here](https://www.notion.so/nuits/Low-Priority-SLURM-Preemption-Notes-and-FAQs-e79f744f9edb47cfa30b81451d4a9d2b)

manasvita commented:
> > the main file needs to be renamed to lowpriority.rst, also should be updated in index.rst: hardware/scavenger -> hardware/lowpriority
> 
> Thanks @rschroter, should be done now.

@ShobanaSekar, you might need to resolve the change request because the Merging is still blocked.

manasvita commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+
+Introduction
+===================
+``lowpriority`` is a new partition on Discovery that allows the research community to use private resources on the HPC cluster when they are idle. 
```
@ShobanaSekar, it might be better to not use "Discovery" and replace that with HPC cluster. So something like:

``lowpriority`` is a new partition on our high performance computing (HPC) cluster that allows the research community to use private resources when they are idle.

ShobanaSekar replied:
@manasvita I've partly addressed this but still think its okay to leave "Discovery" in there, because users might be able to relate to "Discovery" more that "high performance computing". When we get to the stage of migrating things to the next cluster  we can just substitute "Discovery" with the new cluster's name at that point? Let me know if you have strong opinions on this and we can figure it out, thanks!

manasvita replied:
@ShobanaSekar, sure, that sounds good to me.

manasvita commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+===================
+``lowpriority`` is a new partition on Discovery that allows the research community to use private resources on the HPC cluster when they are idle. 
+This new partition has hardware not otherwise available to the general research community and, in time, could double the resources available to NURC users. 
+This is a common practice in HPC clusters to optimize the use of idle private resources that consume power and cooling. 
```
This is a common practice in HPC clusters, which is employed to optimize the use of private resources that consume power via (?) cooling when idle.

I'm not sure what does "consume power and cooling" mean here.

ShobanaSekar replied:
@gshomo  could you clarify here if "power and cooling" is the right usage?

manasvita replied:
@gshomo, let us know what you think and we can resolve this accordingly. Thanks.

dhml replied:
@manasvita, @ShobanaSekar, @gshomo, perhaps just omit "that consume power and cooling"?

manasvita replied:
I agree with @dhml. So maybe just keep it as: 

This is a common practice in HPC clusters to optimize the use of idle private resources

manasvita commented on docs/source/hardware/lowpriorityFAQs.rst:
```diff
@@ -0,0 +1,7 @@
+Low priority partition FAQs
+**********
+
+Or the FAQs can go here as another item in the main menu 
```
Yes, as another item in the main menu is better, so as a new page.

ShobanaSekar replied:
Done

manasvita commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+What is the downside
+===================
+
+Jobs running on the lowpriority partition always carry the risk of being suspended before their wall time ends if a 
```
Jobs running on the lowpriority partition always carry the risk of being suspended before their wall time ends if a request to utilize resources on that partition were to come through from a job deemed as a "high priority" job. See (add LowPriority FAQ link here) for a description of "low" and "high" priority jobs.

manasvita replied:
@ShobanaSekar, this rephrasing looks good. Thanks for taking care of this.

manasvita commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+the lowpriority partition and a high priority job comes through that requires resources currently occupied by the 
+low priority job, then that low priority job will be stopped/suspended within 15s and re-queued. If you have 
+`checkpointing <https://rc-docs.northeastern.edu/en/latest/best-practices/checkpointing.html>`_, implemented in your 
+workflow, such abrupt suspension of jobs would not be an issue. If your job gets killed this way, it’s restart time 
```
such abrupt suspension of jobs would not be an issue -> such as abrupt suspension of the job would not be an issue

ShobanaSekar replied:
Done

manasvita commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+the lowpriority partition and a high priority job comes through that requires resources currently occupied by the 
+low priority job, then that low priority job will be stopped/suspended within 15s and re-queued. If you have 
+`checkpointing <https://rc-docs.northeastern.edu/en/latest/best-practices/checkpointing.html>`_, implemented in your 
+workflow, such abrupt suspension of jobs would not be an issue. If your job gets killed this way, it’s restart time 
```
If your job gets killed this way, it’s restart time depends on the availability of resources at that time. -> However, if your job is stopped/suspended in this way, then it's restart time will depend on the availability of resources at that time.

ShobanaSekar replied:
Done

manasvita commented:
@ShobanaSekar, once you address the above comments then it would be good to run this version by @tketchem14 for language-related edits and approval. Thanks.

marianalevineu commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+
+From most to least-recommended scenarios:
+
+1. Code that can be checkpointed
```
and maybe add the corresponding RC links here as well? unless they are already mentioned elsewhere here.

marianalevineu reviewed (changes requested):
I agree with the comments left by @manasvita

ShobanaSekar commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+1. Code that can be checkpointed
+2. Jobs that fit on a single node
+3. Jobs that require multiple nodes (eg, MPI)
+4. When waiting is too hard
```
@manasvita Did you have an example in mind? I've changed this to "When waiting for your jobs to get started is too hard" for now, but can add example if absolutely needed. I think this is also covered in one of the FAQs (When should I NOT use the lowpriority partition to run my jobs?)

ShobanaSekar commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+Introduction
+===================
+``lowpriority`` is a new partition on Discovery that allows the research community to use private resources on the HPC cluster when they are idle. 
+This new partition has hardware not otherwise available to the general research community and, in time, could double the resources available to NURC users. 
```
Done

ShobanaSekar commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+  srun: job 23498584 queued and waiting for resources
+  srun: job 23498584 has been allocated resources
+
+  $ squeue -u m.joshi
```
Done

ShobanaSekar commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+  $ squeue -u m.joshi
+  JOBID      PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
+  23498592        ctbp     bash  m.joshi  R       0:07      1 d3110
+  23498584 lowpriority     bash  m.joshi  R       2:19      1 d3110
```
Done

ShobanaSekar commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+
+  $ squeue -u m.joshi
+  JOBID      PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
+  23498592        ctbp     bash  m.joshi  R       0:07      1 d3110
```
Done

ShobanaSekar commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+  23498592        ctbp     bash  m.joshi  R       0:07      1 d3110
+  23498584 lowpriority     bash  m.joshi  R       2:19      1 d3110
+
+For the example provided on our `checkpointing <https://rc-docs.northeastern.edu/en/latest/best-practices/checkpointing.html?highlight=array#gromacs-checkpointing-example>`_ page, you can use the low priority partition as::
```
Done

ShobanaSekar commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,79 @@
+===================
+
+To use the low priority partition, just add ``lowpriority`` (no space between 'low' and 'priority') to your partition list. As the partition list is a 
+comma-delimited list of values, it will need to be specified as ``srun --partition=short,lowpriority``. For this example, 
```
Done but with slight modifications as
The partition list is a comma-delimited list of values indicating the order of preference for assigning the job to a particular partition. Hence, the low priority partition...

ShobanaSekar commented:
> @ShobanaSekar, once you address the above comments then it would be good to run this version by @tketchem14 for language-related edits and approval. Thanks.

@manasvita  I've addressed majority of your comments and resolved them. For few of them I've left some comments and kept them unresolved to keep the conversation going. Thanks for all the detailed comments, appreciate it! :)

manasvita commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,75 @@
+To use the low priority partition, just add ``lowpriority`` (no space between 'low' and 'priority') to your partition list. 
+The partition list is a comma-delimited list of values indicating the order of preference for assigning the job to a particular partition. 
+Hence, the low priority partition must be specified after a general partition has been mentioned, as ``srun --partition=short,lowpriority``. 
+For the below example, we will exclude partition ``short`` and focus on ``lowpriority``. When you run the ``squeue`` command, you can see 
```
For the below example -> For the example below

ShobanaSekar replied:
done

manasvita commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,75 @@
+
+  $ squeue -u $USER
+  JOBID      PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
+  23498584 lowpriority     bash  m.joshi  R       2:19      1 d3110
```
m.joshi -> $USER

ShobanaSekar replied:
done

manasvita reviewed (changes requested):
@ShobanaSekar , just added two minor comments

ShobanaSekar commented:
> @ShobanaSekar , just added two minor comments

Thanks @manasvita , those should be taken care of. If this version looks good, we should be able to release this when we are ready, or at least merge with the FAQ branch.

dhml commented:
@manasvita, @marianalevineu, @ShobanaSekar, just a reminder that #26 must be resolved (either merged on closed) before this is merged (I believe this is the last one).

tketchem14 commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,75 @@
+
+Introduction
+===================
+``lowpriority`` is a new partition on our high performance computing (HPC) cluster - Discovery, that allows the research community to use private resources on the HPC cluster when they are idle. 
```
I think we should remove the reference to Discovery, just to save us from having to update when we switch to Explorer. So, it could just read, ". . . on our high performance computing (HPC) cluster that allows the research community. . ."

If you want to keep the reference to Discovery, I suggest taking out the hyphen between cluster and Discovery, replacing it with a comma, and changing 'that' to 'which'.

ShobanaSekar replied:
Thanks @tketchem14 , we did have a conversation about this earlier but I think a lot of our users will actually be more familiar with/relate to "Discovery" than with "high performance computing" so it might be okay to leave it there. We can certainly replace this when we are closer to bringing up the new cluster. Let me know if this doesn't sound okay to you and we can discuss further.

manasvita replied:
@ShobanaSekar , can you pull out the hyphen between cluster and Discovery to address @tketchem14's comment and mark this as resolved? Thanks.

tketchem14 commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,75 @@
+===================
+``lowpriority`` is a new partition on our high performance computing (HPC) cluster - Discovery, that allows the research community to use private resources on the HPC cluster when they are idle. 
+This new partition has hardware that is not otherwise available to the general research community and, in time, could double the resources available to NURC users. 
+This is a common practice in HPC clusters to optimize the use of idle private resources that consume power and cooling. 
```
This paragraph might flow better if we move the third sentence before the second.

ShobanaSekar replied:
Done

tketchem14 commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,75 @@
+
+1. Code that can be `checkpointed <https://rc-docs.northeastern.edu/en/latest/best-practices/checkpointing.html?highlight=array#gromacs-checkpointing-example>`_
+2. Jobs that fit on a single node
+3. Jobs that require multiple nodes (eg, `MPI <https://northeastern-university-rc-public-documentation--19.com.readthedocs.build/en/19/software/mpi.html>`_)
```
the 'eg,' should be changed to, 'e.g., '

ShobanaSekar replied:
Done

tketchem14 commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,75 @@
+for a description of "low" and "high" priority jobs). The members of a PI's partition have priority access to the resources of that partition. 
+This means that if a job is submitted to the ``lowpriority`` partition and a high priority job comes through that requires resources currently occupied by the 
+low priority job, then that low priority job will be stopped/suspended within 30s and re-queued. If you have 
+`checkpointing <https://rc-docs.northeastern.edu/en/latest/best-practices/checkpointing.html>`_, implemented in your 
```
Remove the comma after 'checkpointing.'

ShobanaSekar replied:
Done

ShobanaSekar commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -0,0 +1,75 @@
+This means that if a job is submitted to the ``lowpriority`` partition and a high priority job comes through that requires resources currently occupied by the 
+low priority job, then that low priority job will be stopped/suspended within 30s and re-queued. If you have 
+`checkpointing <https://rc-docs.northeastern.edu/en/latest/best-practices/checkpointing.html>`_, implemented in your 
+workflow, such an abrupt suspension of the job would not be an issue. However, if your job gets suspended this way, then it’s restart time 
```
Done

manasvita commented:
@ShobanaSekar, merging is still blocked due to some unresolved conversations.

manasvita commented on docs/source/hardware/lowpriority.rst:
```diff
@@ -9,9 +9,9 @@ Low priority partition
 ===================
-``lowpriority`` is a new partition on our high performance computing (HPC) cluster - Discovery, that allows the research community to use private resources on the HPC cluster when they are idle. 
-This new partition has hardware that is not otherwise available to the general research community and, in time, could double the resources available to NURC users. 
+``lowpriority`` is a new partition on our high performance computing (HPC) cluster, Discovery which allows the research community to use private resources on the HPC cluster when they are idle. 
```
cluster, Discovery which -> cluster, Discovery, which

joseph-cooke commented:
I am closing this pull request because https://github.com/northeastern-rc/rc-public-documentation/pull/218 is the same material in the new RTD format and with comments addressed. Thank you!!
